### DISCUSSION TITLE
Use Embedded Font on Linux

### DISCUSSION THREAD

**Discussion by tisis2** [AvaloniaUI/Avalonia#12555] [Q&A] [answered]
Hi there,

im currently really struggling with embedded fonts. i am writing a multi lingual application at the moment and need fonts with chinese and korean characters. so i decided to use noto fonts and embed them into the application. then i tried to add a font collection to the fontmanager to make them usable: 
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/dc43c1a3-94ec-4982-9f1c-c2c6d1c25f66)
and use them like 
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/a1bdda7a-da87-4e52-afee-f2640a34cb0f)

but this only resulted in the following behavior
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/f046a169-4350-4a0b-baea-ddabaa833ed1)

then i tried to do 
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/e92fdc92-5216-450e-98ed-71272ca26e31)
 but that also did not help and resulted in the same
 
 can anyone help me how i would correctly embed the fonts so they would also work on linux systems

**Comment by tisis2** [comment]
i reduced to example to only use standard noto sans and korean to reduce the size and be able to attach the example [Fonts_reduced.zip](https://github.com/AvaloniaUI/Avalonia/files/12356600/Fonts_reduced.zip)

**Comment by timunie** [comment]
#11822 

I think you need to make sure that font name matches exactly.

**Reply by tisis2** [reply]
wont it be the same then on windows and linux? or is there a difference how the font lookup works on the different systems? on windows the seem to be found otherwise they would look different. so the names should be correct i think:

with no font family set: 
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/e05328bc-bf26-40b8-9f66-685769f0f8f9)
with font families set as above:
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/202b23ae-d585-47fb-935b-86ed31398d9f)

**Reply by timunie** [reply]
Probably URI lookup on Linux is a bit different? I can only test on windows with your sample and see in DevTools [F12] if the font is the exact correct one.

**Reply by timunie** [reply]
Ha, doesn't work on my Windows machine: 
![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/d0e0c587-d49f-4628-81b8-43772260d19e)

so I guess you have just installed that font, so it finds it in the system

**Reply by tisis2** [reply]
yes you are right. there was still one variant installed. explaining why it was always bold in korean

**Comment by timunie** [comment]
You URI wasn't okay: 

```cs
.ConfigureFonts(fm => fm.AddFontCollection(
    new EmbeddedFontCollection(new Uri("fonts:Noto#Noto Sans, fonts:Noto#Noto Sans KR", UriKind.Absolute),
        new Uri("avares://Assets/Fonts", UriKind.Absolute))))
```

**Reply by Gillibald** [reply]
The first part is the key of the collection. To reference a collection you need to use the fonts scheme `FontFamily="fonts:Noto#Noto Sans"` etc.

**Reply by tisis2** [reply]
i did take a look at the InterFontCollection and did it just the same 
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/b8ed0a84-b46e-465a-996e-e7dcfbdba3cf)

there is no #Inter in the key URI

i tried it like that but already the designer tells me nope

**Reply by timunie** [reply]
`fonts:Noto` works. `#Noto Sans` specifies which variant of `Noto` to use.

**Reply by tisis2** [reply]
so it should theoretically work like that?
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/e53d6970-11f6-4ed9-be6a-9c7b8a0bf5b5)

**Reply by tisis2** [reply]
it seems the windows one falls back to system because it cant find noto. but works if there is only one noto font set in xaml

**Reply by timunie** [reply]
Second URI is still not correct. Please double check. There is not folder Fonts > Assets > Fonts. Its Asstes > Fonts.

**Reply by tisis2** [reply]
the first "Fonts" is the assembly name according to https://docs.avaloniaui.net/docs/getting-started/assets#referencing-assets

i removed it to try but did not change anything:
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/9fc1ef2a-7ead-40c2-a856-e61001b5ba7d)

**Reply by tisis2** [reply]
if i just remove one of the font families i see this exception when i start:
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/9d512a68-1a49-4c05-acaf-fe55b4958348)

but with "avares://Fonts/Assets/Fonts" it starts and the korean font is loaded
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/cda55d81-9e43-4531-9489-c3cdf3b78531)

just does not work anymore as soon as there is a second font family that should be used as fallback
for me it seems the URI is correct so far. but the issue is how the multiple fonts should be referenced

**Reply by Gillibald** [reply]
It is currently impossible to define fallbacks in combination with font collections etc.

But you can define Unicode ranges with FontManagerOptions to achieve this

https://stackoverflow.com/questions/56310609/what-the-chinese-japanese-and-korean-characters-are-in-unicode

https://github.com/AvaloniaUI/Avalonia/pull/7089

**Reply by Gillibald** [reply]
Make sure to add the collection in advance

```
.With(new FontManagerOptions
{
    FontFallbacks = new[]
    {
        new FontFallback
        {
            FontFamily = new FontFamily("fonts:Noto#Noto Sans KR"),
            UnicodeRange = UnicodeRange.Parse("U+20-40,
U+5B-5F,
U+7B-7D,
U+A0-A1,
U+A7,
U+A9,
U+B6-B7,
U+BF,
U+2010-2011,
U+2014-2015,
U+2018-2019,
U+201C-201D,
U+2020-2021,
U+2025-2026,
U+2030,
U+2032-2033,
U+203B,
U+203E,
U+20AC,
U+3001-3003,
U+3008-3011,
U+3014-3015,
U+301C,
U+30FB,
U+3131,
U+3134,
U+3137,
U+3139,
U+3141-3142,
U+3145,
U+3147-3148,
U+314A-314E,
U+AC00-D7A3,
U+FF01-FF03,
U+FF05-FF0A,
U+FF0C-FF0F,
U+FF1A-FF1B,
U+FF1F-FF20,
U+FF3B-FF3D,
U+FF3F,
U+FF5B,
U+FF5D")
        }
    }
 }
```

**Reply by tisis2** [reply]
oh so i have to define application-wide fallbacks...
do i really need to know all the unicode ranges my font would not have the symbols for?
somehow it feels like it breaks the flow of doing such things in xaml. how would i solve the use case when i need a font with fallback for characters in a subcontrol of the application?

**Reply by tisis2** [reply, marked as answer]
i got my usecase to work with the following:
![image](https://github.com/AvaloniaUI/Avalonia/assets/18416393/7f2d0c29-4213-414c-a9d1-0e13415a7a4d)

i just did not define the unicode ranges and it seems to work.
as long as no one requests another font in a part of the application it should work for now.

**Reply by Gillibald** [reply]
Custom fallbacks are used before system fallbacks so it should always work. Unicode ranges are only needed if you want some font for a specific range. The default covers all characters.

**Reply by tisis2** [reply]
so lets say for any reason you need to set a serif font in a control of your application. e.g. an editor inside of your application... i would pick noto serif in addition to noto sans to the fontcollection. noto sans is my application font set in FontManagerOptions with fallbacks for the asian characters.
my editor now can only use serif font for languages that that would not need the asian charset, because i cant set a fallback for the editor control that would use the serif fonts in the editor.
i think that is currently an issue in avalonia. you should be able to set fontfamily with fallbacks in xaml als for embedded fonts not only for system fonts. independent of if its only a font name a fontcollection reference or a uri. as a user a cant understand why there is that difference ATM
